Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,160 @@
+@[to_additive]
+class Quantale (α : Type*) [Semigroup α] extends CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+/-- An integral (or strictly two-sided) additive quantale is a quantale over an additive monoid
+    where top and unit coincide. -/
+class IsIntegralAddQuantale (α : Type*) [AddMonoid α] [AddQuantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_zero : (⊤ : α) = 0
+
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+    top and unit coincide. -/
+@[to_additive]
+class IsIntegralQuantale (α : Type*) [Monoid α] [Quantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_one : (⊤ : α) = 1
+
+section Quantale
+
+variable (α : Type _)
+variable [Semigroup α] [Quantale α]
+
+@[to_additive]
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+
+@[to_additive]
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
+
+end Quantale
+
+section IsIntegral
+open Quantale
+
+variable (α : Type _)
+variable [Monoid α] [Quantale α] [IsIntegralQuantale α]
+
+@[to_additive]
+theorem top_eq_one : (⊤ : α) = 1 := IsIntegralQuantale.top_eq_one
+
+end IsIntegral
+
+namespace Quantale
+
+variable {α : Type _}
+variable [Semigroup α] [Quantale α]
+
+/-- Left- and right- residuation operators on an additive quantale are similar to the Heyting
+operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+@[to_additive "Left- and right- residuation operators on an additive quantale are similar to
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`.
+"]
+def left_residuation (x y : α) := sSup { z | z * x ≤ y }
+
+/-- Left- and right- residuation operators on an additive quantale are similar to the Heyting
+operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ᵣ y = sSup { z | x * z ≤ y }`.
+-/
+@[to_additive "Left- and right- residuation operators on an additive quantale are similar to
+the Heyting operator on complete lattices, but for a non-commutative logic.
+I.e. `x ⇨ᵣ y = sSup { z | x + z ≤ y }`.
+"]
+def right_residuation (x y : α) := sSup { z | x * z ≤ y }
+
+/-- Notation for left-residuation in quantales.
+    I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+@[to_additive "Notation for left-residuation in quantales.
+    I.e. `x ⇨ₗ y = sSup { z | z + x ≤ y }`. "]
+scoped infixr:60 " ⇨ₗ " => left_residuation
```

Comment:
This docstring is what people will see when they hover over your notation, so it's not really useful to tell them it's notation. A good shout is to use `inherit_doc` to automagically copy over the docstring of the definition.
```suggestion
@[to_additive (attr := inherit_doc)]
scoped infixr:60 " ⇨ₗ " => left_residuation
```

Reply:
This does not seem to work entirely. However, the problem seems to be with the use of to_additive on the introduction of notation. When I test the notation on an AddQuantale, it does not seem to have been properly defined while it is for a Quantale. I've updated the code to work properly now by duplicating the notation introduction (once in namespace Quantale, once in namespace AddQuantale) and used the inherit_doc as suggested.